Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -379,9 +379,41 @@ public virtual OrderEvent MarketOnOpenFill(Security asset, MarketOnOpenOrder ord
             // have large gaps, in which case the currentBar.EndTime will be in the past
             // ASUR  | | |      [order]        | | | | | | |
             //  SPY  | | | | | | | | | | | | | | | | | | | |
-            var currentBar = asset.GetLastData();
+            var endTime = DateTime.MinValue;
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                // Get the first valid (non-zero) tick of trade type from an open market
+                var trade = asset.Cache.GetAll<Tick>().FirstOrDefault(x =>
+                    x.TickType == TickType.Trade && x.Price > 0 && asset.Exchange.DateTimeIsOpen(x.Time));
+                if (trade != null)
+                {
+                    endTime = trade.EndTime;
+                    fill.FillPrice = trade.Price;
+                }
+            }
+            else if (subscribedTypes.Contains(typeof(TradeBar)))
+            {
+                var tradeBar = asset.Cache.GetData<TradeBar>();
+                if (tradeBar != null)
+                {
+                    // If the first data after the market opens is a quote bar (high resolution only),
+                    // the open of the last trade bar is not the open for the current day, but the last open in cache.
+                    // We need to verify whether the trade data is from the open market.
+                    if (tradeBar.Period < Resolution.Hour.ToTimeSpan() && !asset.Exchange.DateTimeIsOpen(tradeBar.Time))
+                    {
+                        return fill;
+                    }
+
+                    endTime = tradeBar.EndTime;
+                    fill.FillPrice = tradeBar.Open;
+                }
+            }
+
             var localOrderTime = order.Time.ConvertFromUtc(asset.Exchange.TimeZone);
-            if (currentBar == null || localOrderTime >= currentBar.EndTime) return fill;
+            if (localOrderTime >= endTime) return fill;
 
             // if the MOO was submitted during market the previous day, wait for a day to turn over
             if (asset.Exchange.DateTimeIsOpen(localOrderTime) && localOrderTime.Date == asset.LocalTime.Date)
```

Comment:
What if `TradeBar` was resolution Hour and Daily, check on 405 would be skipped, asset local time could already be market open of the next day by the trade data would be of the previous? 🤔